Stratis.Bitcoin/BlockPulling/LookaheadBlockPuller.cs

```diff
@@ -10,299 +10,400 @@
+        private AutoResetEvent consumed = new AutoResetEvent(false);
+        /// <summary>Event that signals when a new block is pushed to the list of downloaded blocks.</summary>
+        private AutoResetEvent pushed = new AutoResetEvent(false);
+
+        /// <summary>Identifies the last block that is currently being requested/downloaded.</summary>
+        private ChainedBlock lookaheadLocation;
+        /// <summary>Identifies the last block that is currently being requested/downloaded.</summary>
+        public ChainedBlock LookaheadLocation
+        {
+            get
+            {
+                return this.lookaheadLocation;
+            }
+        }
+
+        /// <summary>Median of a list of past downloadedCounts values. This is used just for logging purposes.</summary>
+        /// <remarks>TODO: There is a race condition that could cause GetMedian method to be called with empty downloadCounts list and throw exception.</remarks>
+        public decimal MedianDownloadCount
+        {
+            get
+            {
+                if (this.downloadedCounts.Count == 0)
+                    return decimal.One;
+                return (decimal)GetMedian(this.downloadedCounts);
+            }
+        }
+
+        /// <summary>If true, the puller is a bottleneck.</summary>
+        public bool IsStalling
+        {
+            get;
+            internal set;
+        }
+
+        /// <summary>If true, the puller consumer is a bottleneck.</summary>
+        public bool IsFull
+        {
+            get;
+            internal set;
+        }
+
+        /// <inheritdoc />
+        public void SetLocation(ChainedBlock tip)
+        {
+            Guard.NotNull(tip, nameof(tip));
+            this.location = tip;
+        }
+
+        /// <inheritdoc />
+        public void RequestOptions(TransactionOptions transactionOptions)
+        {
+            if (transactionOptions == TransactionOptions.Witness)
+            {
+                this.Requirements.RequiredServices |= NodeServices.NODE_WITNESS;
+                foreach (var node in this.Nodes.Select(n => n.Behaviors.Find<BlockPullerBehavior>()))
+                {
+                    if (!this.Requirements.Check(node.AttachedNode.PeerVersion))
+                    {
+                        node.ReleaseAll();
+                    }
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public Block NextBlock(CancellationToken cancellationToken)
+        {
+            this.downloadedCounts.Add(this.DownloadedBlocks.Count);
+            if (this.lookaheadLocation == null)
+            {
+                // Calling twice is intentional here.
+                // lookaheadLocation is null only during initialization
+                // or when reorganisation happens. Calling this twice will 
+                // make sure the initial work of the puller is away from 
+                // the lower boundary.
+                AskBlocks();
```

Comment:
LookaheadBlockPuller makes sure that:

```
if B - A <= ActualLookahead then AskBlocks() get called.
```

At initialization B == A so we need to call AskBlocks() two times.